common/membership/ringpop.go

```diff
@@ -87,12 +88,13 @@ func (r *RingPop) bootstrap(
 	bootstrapHostPostRetriever func() ([]string, error),
 	bootstrapRetryBackoffInterval time.Duration,
 ) {
-	retryCount := 0
-
-	for {
+	policy := backoff.NewExponentialRetryPolicy(bootstrapRetryBackoffInterval)
+	policy.SetBackoffCoefficient(1)
+	policy.SetMaximumAttempts(maxBootstrapRetries)
```

Comment:
These settings match the existing behavior, but what I'd really like to do here is set Expiration to `r.maxJoinDuration` and use an actual exponential backoff. Currently the behavior is to wait 5 times `r.maxJoinDuration` before giving up, which doesn't really match expectations.

It would be better to attempt to bootstrap with, say, 5s timeout, retry up to `maxJoinDuration` (default 30s), and reread bootstrap hosts each time. This could recover from some weird situations faster. Right now it only rereads bootstrap hosts every 30s, but we set the fx application startup timeout to 15s, and block on this, so we "give up" before we give it a second chance. (It turns out the startup timeout doesn't really do anything though.)

If we agree that's a good idea, I can either do it in this PR or a following one.

Reply:
Your proposal make sense.